Total points for this HW: 100.

Please note: **Copying and pasting other people's work is absolutely prohibited**. Any such cases will be reported to CUSP's education team and severely punished. Discussion is encouraged, and feel free to exchange ideas with your classmates, but please write your own code and do your own work.


<span style='color:red'>  pts</span>

In [2]:
from IPython.display import Image
import pandas as pd
import numpy as np

### Q1 KNN (35 points)   <span style='color:red'>  pts</span>

Given a dataset with 53,414 points in "HW3Q1.txt" which is KEGG Metabolic Relation Network (Directed) Data Set
https://archive.ics.uci.edu/ml/datasets/KEGG+Metabolic+Relation+Network+%28Directed%29, you should ignore the first column which is not real value and focus on columns 2 to 24.

a) implement the function of Euclidean distance, test it on the first two points in the given dataset, and output the distance. (10 points)

Attribute Information:

a) Pathway text   
b) Nodes integer (min:2, max:116)   
c) Edges integer (min:1, max:606)   
d) Connected Components integer (min:1, max:13)     
e) Network Diameter integer (min:1, max:30)   
f) Network Radius integer (min:1, max:2)    
g) Shortest Path integer (min:1, max:3277)    
h) Characteristic Path Length real (min:1, [Web Link])    
i) Avg.num.Neighbours real (min:1, [Web Link])    
j) Isolated Nodes integer (min:0, max:1)     
k) Number of Self Loops integer (min:0, max:0)    
l) Multi-edge Node Pair integer (min:0, max:57)     
m) NeighborhoodConnectivity real (min:1, [Web Link])    
n) Outdegree real (min:0.5, [Web Link])     
o) Stress real (min:0, [Web Link])      
p) SelfLoops integer (min:0, max:0)     
q) PartnerOfMultiEdgedNodePairs real (min:0, [Web Link])    
r) EdgeCount real (min:1, [Web Link])      
s) BetweennessCentrality real (min:0, [Web Link])     
t) Indegree real (min:0.5, [Web Link])        
u) Eccentricity real ([Web Link], [Web Link])     
v) ClosenessCentrality real ([Web Link], max:1)     
w) AverageShortestPathLength real ([Web Link], [Web Link])     
x) ClusteringCoefficient real (min:0, [Web Link])     

In [6]:
# your answers here
q1=pd.read_csv('HW3Q1.txt', header=None)
q1.head()

0   1   2   3   4   5    6         7         8   9   ...         14  \
0  aac00010  26  43   2   7   1  211  3.222749  3.230769   0  ...  27.000000   
1  aac00020  27  52   2  10   1  226  3.411504  3.851852   0  ...  26.629630   
2  aac00030  26  53   2   6   1  202  2.732673  3.769231   0  ...  14.961538   
3  aac00040  20  28   1   8   1   65  2.523077  2.600000   0  ...   4.950000   
4  aac00051  15  33   1   4   2   85  1.858824  4.400000   0  ...   5.000000   

   15        16        17        18        19        20        21        22  \
0   0  0.076923  3.307692  0.062418  1.653846  3.923077  0.375199  2.210114   
1   0  0.000000  3.851852  0.036567  1.925926  4.074074  0.399880  2.168687   
2   0  0.230769  4.076923  0.026604  2.038462  3.115385  0.430368  1.864288   
3   0  0.100000  2.800000  0.014474  1.400000  2.450000  0.481044  1.492500   
4   0  0.000000  4.400000  0.026740  2.200000  2.333333  0.518691  1.532326   

         23  
0  0.221154  
1  0.290197  
2  0.230311  
3  0.210000  
4  0.267222  

[5 rows x 24 columns]

In [54]:
def euc_distance(a,b,df):
    diff = 0
    for i in range(1,df.shape[1]):
        x=a[i].reset_index(drop=True)
        y=b[i].reset_index(drop=True)
        diff += np.square(x-y)
    return np.sqrt(diff).to_numpy()[0]

In [55]:
euc_distance(q1.iloc[:1],q1.iloc[1:2],q1)

17.837937709396144

b) given the first point as query, implement the baseline (slide 10 of lecture 8) which scans every point in the rest of the file, return the nearest neighbor to the query. In the end, output the resulted point and the distance to the query. (10 points)

In [57]:
# your answers here
import time
start_time = time.time()
distances=[]
for i in range(1,q1.shape[0]):
    distances.append(euc_distance(q1.iloc[:1],q1.iloc[i:i+1],q1))
print("--- %s seconds ---" % (time.time() - start_time))

In [60]:
len(distances)

53412

In [59]:
np.min(distances)

12.466167487557817

In [61]:
distances.index(min(distances))

45833

In [64]:
euc_distance(q1.iloc[:1],q1.iloc[45834:45834+1],q1)

12.466167487557817

In [68]:
q1.iloc[45834:45834+1][0]

45834    spj00550
Name: 0, dtype: object

The nearest neighbor is spj00550. The minimum distance is 12.466167487557817.

c) implement the index acceleration using Ball-tree, and present the nearest neighbor of the first point (to call the library of Ball-tree, refer to: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.BallTree.html) (10 points)

In [70]:
q1.iloc[:,1:]

1   2   3   4   5    6         7         8   9   10  ...         14  \
0      26  43   2   7   1  211  3.222749  3.230769   0   0  ...  27.000000   
1      27  52   2  10   1  226  3.411504  3.851852   0   0  ...  26.629630   
2      26  53   2   6   1  202  2.732673  3.769231   0   0  ...  14.961538   
3      20  28   1   8   1   65  2.523077  2.600000   0   0  ...   4.950000   
4      15  33   1   4   2   85  1.858824  4.400000   0   0  ...   5.000000   
...    ..  ..  ..  ..  ..  ...       ...       ...  ..  ..  ...        ...   
53408  20  25   2   5   1   62  2.112903  2.500000   0   0  ...   5.550000   
53409   4   3   1   3   1    6  1.666667  1.500000   0   0  ...   1.000000   
53410   5   6   1   3   1   10  1.500000  2.400000   0   0  ...   1.000000   
53411  16  24   3   4   1   39  1.666667  2.250000   0   0  ...   2.250000   
53412  11  10   2   4   1   21  1.857143  1.818182   0   0  ...   1.636364   

       15        16        17        18        19        20        21  \
0       0  0.076923  3.307692  0.062418  1.653846  3.923077  0.375199   
1       0  0.000000  3.851852  0.036567  1.925926  4.074074  0.399880   
2       0  0.230769  4.076923  0.026604  2.038462  3.115385  0.430368   
3       0  0.100000  2.800000  0.014474  1.400000  2.450000  0.481044   
4       0  0.000000  4.400000  0.026740  2.200000  2.333333  0.518691   
...    ..       ...       ...       ...       ...       ...       ...   
53408   0  0.000000  2.500000  0.043910  1.250000  1.650000  0.377880   
53409   0  0.000000  1.500000  0.166667  0.750000  1.500000  0.541667   
53410   0  0.000000  2.400000  0.083333  1.200000  1.600000  0.593333   
53411   0  0.250000  2.500000  0.018056  1.250000  1.375000  0.454412   
53412   0  0.000000  1.818182  0.029221  0.909091  1.454545  0.468615   

             22        23  
0      2.210114  0.221154  
1      2.168687  0.290197  
2      1.864288  0.230311  
3      1.492500  0.210000  
4      1.532326  0.267222  
...         ...       ...  
53408  1.077955  0.157500  
53409  1.125000  0.000000  
53410  1.150000  0.166667  
53411  0.923438  0.188542  
53412  0.989394  0.068182  

[53413 rows x 23 columns]

In [85]:
# your answers here
from sklearn.neighbors import NearestNeighbors,BallTree
import time
start_time = time.time()
tree = BallTree(q1.iloc[:,1:], leaf_size=40)              
dist, ind = tree.query(q1.iloc[:1,1:], k=2) 
print("--- %s seconds ---" % (time.time() - start_time))

--- 0.35341596603393555 seconds ---


In [86]:
dist[0][1], ind[0][1]

(12.466167487557817, 45899)

d) compare the running time of the above two methods you have implemented (to measure the running time, refer to
https://stackoverflow.com/questions/1557571/how-do-i-get-time-of-a-python-programs-execution) (5 points)

In [ ]:
# your answers here
#The BallTree reduce the time a lot compare with the brute force method.

### Q2. Trajectory Distance Metric and KNN (35 points)  <span style='color:red'>  pts</span>

In this question, we use a dataset with 1114 trajectories in Porto, Portugal: "HW3Q2.csv" for trajectory distance computation and kNN. (To access the full dataset, refer to https://archive.ics.uci.edu/ml/datasets/Taxi+Service+Trajectory+-+Prediction+Challenge,+ECML+PKDD+2015)

a) implement DTW and test it on the first two trajectories in the file, and output the distance. (5 points)

In [109]:
# your answers here
import re
f = open("HW3Q2.txt", "r")
t=[]
point=[]
for x in f:
    ##line = re.findall(r"[-+]?\d*\.\d+|\d+", x)
    line_list = x.split() 
    t.append(line_list[1])
    point.append(line_list[0])
    ##t.append(line)

In [91]:
len(t)

1114

In [101]:
import ast
traj=[]
for i in range(len(t)):
    traj.append(np.array(ast.literal_eval(t[i])))

In [19]:
import numpy as np

def eucl_dist(x, y):
    """
    Usage
    -----
    L2-norm between point x and y
    Parameters
    ----------
    param x : numpy_array
    param y : numpy_array
    Returns
    -------
    dist : float
           L2-norm between x and y
    """
    dist = np.linalg.norm(x - y)
    return dist

######################
# Euclidean Geometry #
######################

def e_dtw(t0, t1):
    """
    Usage
    -----
    The Dynamic-Time Warping distance between trajectory t0 and t1.
    Parameters
    ----------
    param t0 : len(t0)x2 numpy_array
    param t1 : len(t1)x2 numpy_array
    Returns
    -------
    dtw : float
          The Dynamic-Time Warping distance between trajectory t0 and t1
    """

    n0 = len(t0)
    n1 = len(t1)
    C = np.zeros((n0 + 1, n1 + 1))
    C[1:, 0] = float('inf')
    C[0, 1:] = float('inf')
    for i in np.arange(n0) + 1:
        for j in np.arange(n1) + 1:
            C[i, j] = eucl_dist(t0[i - 1], t1[j - 1]) + min(C[i, j - 1], C[i - 1, j - 1], C[i - 1, j])
    dtw = C[n0, n1]
    return dtw

In [104]:
e_dtw(traj[0],traj[1])

1.0735257244071137

b) implement LCSS and test it on the first two trajectories in the file, by setting the threshold as 0.001, and output the similarity score. (5 points)

In [105]:
# your answers here
def e_lcss(t0, t1, eps):
    """
    Usage
    -----
    The Longuest-Common-Subsequence distance between trajectory t0 and t1.
    Parameters
    ----------
    param t0 : len(t0)x2 numpy_array
    param t1 : len(t1)x2 numpy_array
    eps : float
    Returns
    -------
    lcss : float
           The Longuest-Common-Subsequence distance between trajectory t0 and t1
    """
    n0 = len(t0)
    n1 = len(t1)
    # An (m+1) times (n+1) matrix
    C = [[0] * (n1 + 1) for _ in range(n0 + 1)]
    for i in range(1, n0 + 1):
        for j in range(1, n1 + 1):
            if eucl_dist(t0[i - 1], t1[j - 1]) < eps:
                C[i][j] = C[i - 1][j - 1] + 1
            else:
                C[i][j] = max(C[i][j - 1], C[i - 1][j])
    lcss = 1 - float(C[n0][n1]) / min([n0, n1])
    return lcss

In [106]:
e_lcss(traj[0],traj[1],0.001)

1.0

c) implement kNN of using a DTW and lCSS and test it using the first trajectory as query and the rest 1113 trajectories as the dataset. (10 points)

In [118]:
# your answers here
k_dtw={}
def dtw_kNN(k):
    for i in range(1,len(traj)):
        dtw=e_dtw(traj[0],traj[i])
        k_dtw[point[i]] = dtw
    sort_dtw=sorted(k_dtw.items(), key=lambda d: d[1])
    return sort_dtw[:k]

In [124]:
#test by k=1
dtw_kNN(1)

[('860', 0.12132758515143127)]

In [121]:
k_lcss={}
def lcss_kNN(k,eps):
    for i in range(1,len(traj)):
        lcss=e_lcss(traj[0],traj[i],eps)
        k_lcss[point[i]] = lcss
    sort_lcss=sorted(k_lcss.items(), key=lambda d: d[1])
    return sort_lcss[:k]

In [125]:
#test by k=1
lcss_kNN(1,0.001)

[('810', 0.3157894736842105)]

d) compare the running time of two approaches by using the first trajectory as query, and check whether them return the same results by calling 5NN. (5 points)

In [127]:
# your answers here
import time
start_time = time.time()
dtw_kNN(5)

[('860', 0.12132758515143127),
 ('551', 0.12909118284621376),
 ('440', 0.1487935698018854),
 ('90', 0.16888369178488338),
 ('284', 0.20669792029804407)]

In [128]:
print("--- %s seconds ---" % (time.time() - start_time))

--- 10.421721935272217 seconds ---


In [130]:
import time
start_time = time.time()
lcss_kNN(5,0.001)

[('810', 0.3157894736842105),
 ('963', 0.368421052631579),
 ('896', 0.4),
 ('442', 0.4736842105263158),
 ('853', 0.4736842105263158)]

In [131]:
print("--- %s seconds ---" % (time.time() - start_time))

--- 7.3508689403533936 seconds ---


lcss is faster than dtw, and they didn't return the same result.

e) visualize the query trajectory, and its 1NN and 5NN with LCSS using Scikit-learn mobility (https://github.com/scikit-mobility/scikit-mobility) (10 points)

In [133]:
import skmob

ModuleNotFoundError: No module named 'skmob'

In [ ]:
# your answers here


### Q3. Clustering: Spatial and Temporal Distributions of Chicago Crimes (30 points)  <span style='color:red'>  pts</span>

In this question you will use k-means clustering in sklearn to answer the question, “Do different types of crime display different trends over space and time?”  The dataset “HW3Q3_1.csv” consists of data for 119 different types of crime, each of which occurred at least 100 times in Chicago during the year 2016.  For each crime type, we have various features representing the spatial and temporal distribution of crime, including:

- The proportion of all crimes of that type that occurred on each day of the week (day_Sun, day_Mon, …, day_Sat).

- The proportion of all crimes of that type that occurred on each hour of the day (hour_0 = midnight to 12:59am, hour_1 = 1am to 1:59am, …, hour_23 = 11pm to 11:59pm).

- The proportion of all crime of that type that occurred in each of the 77 community areas of Chicago (community_area_1 … community_area_77).

We also have, for each crime type, its categorization by the FBI:

- Category = “P1V” corresponds to Part 1 Violent Crime, i.e., serious violent crimes

- Category = “P1P” corresponds to Part 1 Property Crime, i.e., serious property crimes

- Category = “P2” corresponds to Part 2 (less serious) crimes.

To answer parts a through f, you should cluster the 119 crime types using k-means into k = 3 clusters using only the hour of day (hour_0..hour_23) attributes.

a) Copy each cluster’s mean values for hour_0…hour_23 into a DataFrame and create a line graph to visualize these values by cluster.  (5 pts)

b) Describe the three different hour-of-day trends represented by these three clusters (5 pts).

c) Do you notice any consistent trends about which crime types are assigned to which cluster?  Note that by a "crime type", we are referring to specific crimes such as "narcotics" or "assault", not the FBI categories. (5 pts)  

d) Do the three clusters have different day-of-week trends?  Again, visualize the trends for each cluster by creating a line graph and discuss any notable differences.  (5 pts)

e) Do the three clusters affect different types of communities/neighborhoods?  To answer this question, you could first compute the proportions of "cluster 1", "cluster 2", and "cluster 3" crimes for each community area, and identify particular community areas with disproportionate amounts of a given cluster.  You can then use the provided file (HW3Q3_2.csv), to determine whether these community areas have any notable common characteristics (poverty, overcrowding, etc.).  You may also wish to consult the Chicago Community Areas map at https://en.wikipedia.org/wiki/Community_areas_in_Chicago. (5 pts)

f) How well do the three groups formed by clustering hour-of-day trends correspond to the FBI’s division between P1V, P1P, and P2 crimes? (5 pts)



In [ ]:
data3=pd.read_csv("HW3Q3_1.csv")
data3.head()

In [ ]:
data4=pd.read_csv("HW3Q3_2.csv")
data4.head()

In [ ]:
# your answers here